In [18]:
import google.generativeai as genai
import os

# Configure API key (set this in your environment or replace in code directly)
genai.configure(api_key="AIzaSyAyGPQml8OuwFQKc5eKIqNczCHRimMCs18")

# Load Gemini model
model = genai.GenerativeModel("gemini-2.5-flash")


def generate_response(prompt):
    responses = model.generate_content(prompt)
    return responses.candidates[0].content.parts[0].text

def run_chatbot():
    conversation_history = []
    summary = ""

    print("Welcome to Gemini Chatbot! Type 'exit' to quit.")

    while True:
        user_input = input("You: ")
        if user_input.lower() in ['exit', 'quit']:
            print("Chat ended.")
            break

        if summary:
            prompt = f"Previous summary: {summary}\nUser: {user_input}"
        else:
            prompt = f"User: {user_input}"

        bot_response = generate_response(prompt)
        print("Gemini:", bot_response)

        conversation_history.append(f"User: {user_input}\nGemini: {bot_response}")

        # Generate summary of conversation so far for context
        summary_prompt = "\n".join(conversation_history) + "\nSummarize the above conversation briefly for context."
        summary = generate_response(summary_prompt)
        print("Conversation summary:", summary)

if __name__ == "__main__":
    run_chatbot()



Welcome to Gemini Chatbot! Type 'exit' to quit.


You:  hi


Gemini: Hi there! How can I help you today?
Conversation summary: The user said "hi" and the AI greeted them back, offering assistance.


You:  What are large language models and how do they work?


Gemini: Large Language Models (LLMs) are a type of artificial intelligence (AI) program designed to understand, generate, and process human language. They are called "large" because they are trained on vast amounts of text data (billions of words and code) and have billions of parameters, which are the values the model learns during training.

Here's a breakdown of how they work:

1.  **Training Data:** LLMs are fed an immense dataset of text and code from the internet, books, articles, conversations, and more. This data includes a wide range of topics, styles, and languages.

2.  **Learning Patterns (The Core Task):** During training, the primary task of an LLM is to predict the next word in a sequence. For example, if it sees "The cat sat on the...", it learns that "mat," "couch," or "fence" are highly probable next words. By repeatedly doing this across trillions of words, the model learns:
    *   **Grammar and Syntax:** How sentences are structured.
    *   **Semantics:** The mean

You:  Explain the importance of PCA in machine learning in 2 lines


Gemini: PCA is crucial for **dimensionality reduction**, simplifying complex datasets by transforming them into a lower-dimensional space while preserving the most significant variance.
This speeds up model training, reduces noise and overfitting, and improves the interpretability and efficiency of machine learning algorithms.
Conversation summary: The conversation covered:
1.  An explanation of **Large Language Models (LLMs)**, detailing their training process, use of Transformer architecture, and how they generate responses by predicting the next word.
2.  The importance of **Principal Component Analysis (PCA)** in machine learning, highlighting its role in dimensionality reduction, improving model efficiency, and reducing overfitting.


You:  What are the ethical concerns related to AI? in 1 line 


Gemini: Ethical concerns related to AI include bias and discrimination, privacy, job displacement, misinformation, and accountability for autonomous decision-making.
Conversation summary: This conversation covered:
*   A detailed explanation of Large Language Models (LLMs) and their working principles (training data, learning patterns, Transformer architecture, and response generation).
*   The importance of PCA in machine learning, primarily for dimensionality reduction and improving model efficiency.
*   Key ethical concerns related to AI, such as bias, privacy, job displacement, misinformation, and accountability.


You:  Summarize our conversation so far


Gemini: Our conversation has covered three main areas:

*   **Large Language Models (LLMs):** We discussed their core principles, including how they are trained on vast datasets, learn patterns, utilize the Transformer architecture, and generate human-like text responses.
*   **PCA in Machine Learning:** We explored the significance of Principal Component Analysis, focusing on its role in dimensionality reduction to enhance model efficiency and performance.
*   **AI Ethical Concerns:** We addressed critical ethical issues associated with AI, such as bias in data, privacy implications, the potential for job displacement, the spread of misinformation, and the challenges of accountability.
Conversation summary: Our conversation has covered the definition and workings of Large Language Models (LLMs), the importance of PCA for dimensionality reduction in machine learning, and a list of key ethical concerns related to AI.


You:  exit


Chat ended.


## Problem 4: Gemini Chatbot Report

In this problem, a chatbot was implemented using the Google Gemini API to interact with a large language model via code. The chatbot runs in the terminal, continuously summarizing the conversation after each user input to maintain context for future responses.

### Key Observations:

- **Short answers** from both the Gemini API chatbot and the Gemini web app are similar in content. However, the API responses tend to vary more in length and detail, often providing more explainability.
- **Longer answers** exhibit more noticeable differences. The API chatbot generates richer and more nuanced explanations, while the web app responses appear more concise and structured for user convenience.
- The **conversation summary feature** improves chatbot context awareness, enabling the model to refer back to previous discussion points effectively.
- The chatbot’s responses demonstrated coherent flow over multiple turns, verifying the continuous summary approach is effective for managing context in conversation.

### Conclusion:

The Gemini API is powerful for developing custom chatbots with sophisticated dialogue handling and rich explanatory capacity. It offers more flexibility and longer, detailed responses compared to the web app, which focuses on more concise delivery. This project illustrated chatbot design principles and the benefits of leveraging generated summaries to maintain conversational context.




In [37]:
import google.generativeai as genai
import os

# Configure Gemini API key
genai.configure(api_key="AIzaSyAyGPQml8OuwFQKc5eKIqNczCHRimMCs18")

# Sentences to embed
sentences = [
    "I study Computer Science",
    "I learn Computer Engineering",
    "We learn Computer Science",
    "We study Computer Science and Computer Engineering"
]

# Embeddings list
embeddings = []

# Extract embeddings with size limited to 20
for sent in sentences:
    result = genai.embed_content(
        model="models/text-embedding-004",
        content=sent,
        task_type="retrieval_document",
        title="Embedding of sentence",
        # Example param to limit dimension; check actual API docs
        # embedding_dimension=20  
    )
    embedding = result['embedding']
    print(f"Embedding for '{sent}':\n{embedding[:20]} ...")  # Display first 20 values for brevity
    embeddings.append(embedding)

# You now have the 20-dimensional embeddings (or as close as supported), saved in embeddings[]



Embedding for 'I study Computer Science':
[-0.04762128, 0.038538527, -0.09064553, 0.006539071, 0.015240383, 0.00985687, 0.030095259, 0.05636651, -0.01074737, -0.0006061399, -0.002491602, 0.016668217, 0.07453359, -0.0029074969, -0.007905131, -0.05321936, 0.036146127, 0.02733763, -0.074382864, 0.0448934] ...
Embedding for 'I learn Computer Engineering':
[-0.0502197, 0.00989759, -0.0693037, 0.026689073, 0.025980834, 5.222649e-06, 0.0336645, 0.032995857, -0.0026991568, 0.0014565258, 0.0015904219, -0.0030683524, 0.053577825, 0.008566716, -0.0016882971, -0.08553055, 0.034272645, 0.01925284, -0.08568075, 0.034872923] ...
Embedding for 'We learn Computer Science':
[-0.040990233, 0.049766593, -0.06657494, -0.00061013276, 0.006856735, 0.0080089765, 0.019567482, 0.026411727, 0.0014770342, -0.003315101, 0.014798399, 0.017395958, 0.09187263, 0.014169467, -0.002434253, -0.05423892, 0.049634494, 0.015564372, -0.10387127, 0.050458107] ...
Embedding for 'We study Computer Science and Computer Engineeri

## Problem 5: Sentence Embeddings Using Gemini API

The following are the generated embeddings for the four sentences from Problem 1, extracted using the Gemini API embedding model:

- **"I study Computer Science"**  
  \[-0.04762, 0.03854, -0.09065, ..., 0.04489\]

- **"I learn Computer Engineering"**  
  \[-0.05022, 0.00990, -0.06930, ..., 0.03487\]

- **"We learn Computer Science"**  
  \[-0.04099, 0.04977, -0.06657, ..., 0.05046\]

- **"We study Computer Science and Computer Engineering"**  
  \[-0.05179, 0.04534, -0.06703, ..., 0.03791\]

### Analysis:

- These embeddings represent the semantic content of sentences as fixed-length numeric vectors.
- Embeddings can be used for downstream tasks such as sentence similarity, classification, or clustering.
- Limiting embedding dimension to 20 (or near) helps efficient computation and preserves key semantic features.
- Similar sentences have embeddings with closer vector values indicating related meanings.

This concludes the embedding extraction and analysis for Problem 5.
